### ISSUE TITLE
Wrap UniqueSequences(), RemoveAxis() and GetLayer() to python; make sure IntersectDevice() wrapped

### ISSUE THREAD

**Issue by danpovey** [k2-fsa/k2#641]
These need to be available from Python. 

The RemoveAxis() is for `Ragged<int32_t>`.

The way we'll use them is as follows:
```
     # paths will be k2.RaggedInt with 3 axes: [seq][path][arc_pos], containing arc_idx012
     paths = mbr_lats.random_paths(True, 100);
     # phone_seqs will be k2.RaggedInt like paths, but containing phones (and final -1's,
     # and 0's for epsilon)
     phone_seqs = k2.index(mbr_lats.phones, paths)
     # Remove epsilons from `phone_seqs`
     phone_seqs = k2.ragged.remove_values_eq(phones, 0)
     # Remove repeated sequences from `phone_seqs`
     phone_seqs = k2.ragged.unique_sequences(phone_seqs)
     # Remove the 1st axis from `phone_seqs` (that corresponds to `seq`) and
     # keep it for later, we'll be processing paths separately.
     seq_to_path_shape = k2.ragged.get_layer(phone_seqs.shape, 0)
     path_to_seq_map = seq_to_path_shape.row_ids(1)
     phone_seqs = k2.ragged.remove_axis(phone_seqs, 0)


     # now compile decoding graphs corresponding to `phone_seqs` by constructing
     # fsas from them (remember we already have the final -1's!) and composing with ctc_topo.
      paths_decoding_graphs = ...
      # Now extract the sets of paths from the lattices corresponding to each of those n-best
      # phone sequences; these will effectively be lattices with one path but alternative
      # alignments.
      paths_lats = k2.intersect_device(mbr_lats, paths_decoding_graphs, path_to_seq_map)

     # Of course there is plenty more that will happen after we've got these phone
     # sequences.
```

**Comment by csukuangfj** [comment]
Will wrap it tomorrow.

**Comment by csukuangfj** [comment]
How is the resulting `paths_lats` going to be used?

**Comment by danpovey** [comment]
Another note on the next steps after this.  This is where we can start computing confidences per phone.


  - First we'll compute the phone-level posteriors from `num_lats` (MMI numerator) and `paths_lats`, using the `phones` attribute, and getting rid of elements where the phone is 0.  Let these be called `num_phone_post`, indexed (seqframe_idx, phone), and `paths_phone_post`, indexed (pathframe_idx, phone).    seqframe_idx is an idx01 that includes the sequence-index and the frame, i.e. the index into axis 0 of dense_fsa_vec.scores.  pathframe_idx is the same but for paths, I'll show later how to construct it.

  
  - The indexes into `num_phone_post` and `paths_phone_post` are not the same, so we need to expand `num_phone_post` to be indexed by pathframe_idx before subtracting.    I suggest we do it as follows... when we create `mbr_lats` using intersect_dense_pruned() we request the `seqframe_idx`, and this is the same index as the `seqframe_idx` of `num_phone_post`.  This will get propagated from `mbr_lats` into `paths_lats`, but that's not actually the seqframe_idx we want to create the posteriors (since it collapses the posteriors from different paths of the same FSA).   
  
  - The following will allow us to map the seqframe_idx from the FSA-based one to the path-based one.  Let the nnet output as a DenseFsaVec that we use for decoding be called `dense_fsa_vec.`.  Its shape is `dense_fsa_vec.shape`.   Code draft:
 ```
     # by seq we mean the original sequence indexes, by path we mean the indexes
     # of the n-best paths; path_to_seq_map maps from path-index to seq-index.
     seqs_shape = dense_fsa_vec.shape  # we originally decoded with this
                                 # dense_fsa_vec, it's the nnet outputs.
     # paths_shape will also be a k2.RaggedInt with 2 axes.
     paths_shape = k2.index(seqs_shape, path_to_seq_map);
     seq_starts, path_starts = seqs_shape.row_splits(1)[:-1], paths_shape.row_splits(1)[:-1]
     # We can map from seqframe_idx  for paths, to seqframe_idx for seqs, by adding
     # path_offsets.  path_offsets is indexed by path-index.
     path_offsets = path_starts - seq_starts[path_to_seq_map]
     with paths_shape.row_ids(1) as paths_shape_row_ids:
        pathframe_idx_to_seqframe_idx =  torch.arange(0, paths_shape_row_ids.shape[0]) - \
                      path_offsets[paths_shape.row_ids]
     # assign new attribute 'pathframe_idx' that combines path and frame..
     # may need to use k2.index below as torch doesn't support int32_t indexes.
     paths_lats_arc2path = paths_lats.arcs.shape.row_splits(1)[paths_lats.arcs.shape.row_splits(2)]
     paths_lats.pathframe_idx = paths_lats.seqframe_idx + path_offsets[paths_lats_arc2path]

     # now suppose the phone posteriors of the MMI numerator were called num_phone_post,
     # indexed (seqframe_idx, phone).  We create a new variable num_phone_post_paths,
     # which is like num_phone_post but indexed (pathframe_idx, phone):
     # Note: if the following indexing operation, sparse_matrix[1d tensor], does not exist,
     # we may have to create it.  It's indexing axis 0; we could use ragged tensor indexing
     # if num_phone_post was represented as a ragged tensor, which unfortunately it isn't.
     num_phone_post_paths = num_phone_post[pathframe_idx_to_seqframe_idx]

     # errors_per_pathframe_idx would be a tensor of float, indexed by pathframe_idx, containing
     # errors between 0 and 2.
     errors_per_pathframe_idx = (num_phone_post_paths - paths_phone_post).abs().sum(axis=1)
```
So the above code gives us an error metric per pathframe_idx, which can be the basis of our confidence estimation
(actually number-of-errors estimation).  We next want to map this confidence onto a position in the path, as in the index of the phone (not frame).  In fact we want a way to map from frames to phones and vice versa, for purposes of getting the features as well as the targets for this confidence estimation.  We'll be using something called a pathphone_idx, which will combine the path-index and the position in the phone sequence.  In fact it will contain a little more information than that, because we also assign a separate index to the blanks at the beginning and end and between each position in the phone sequence.

The process of creating the decoding FSAs from the phone sequences would be something like this:
```
     phone_fsas = k2.create_linear_fsas(phone_seqs)
     # add self loops
     phone_fsas = k2.add_self_loops(phone_fsas)
     # set an attribute called pathphone_idx, which corresponds to the arc-index in `phone_fsas` with
     # self-loops.  each phone has an index but there are blanks between them and at the start and end.
     phone_fsas.pathphone_idx = torch.arange((0, phone_fsas.arcs.shape[0]), dtype=torch.int32)
```

Later on when we have paths_lats, we'll use the arc posteriors to construct a sparse matrix pathframe_to_pathphone indexed (pathframe_idx, pathphone_idx).  We'll be using this to multiply embeddings and the errors_per_pathframe_idx.

**Comment by danpovey** [comment]
Re your comment "How is the resulting paths_lats going to be used?" .. see the comment above this one.

**Comment by danpovey** [comment]
actually I am a little confused why that comment only showed up  just now, I made it a long time ago and didn't just commit it or anything like that.  Looks like some kind of bug in github.

**Comment by csukuangfj** [comment]
I have several doubts about the above draft code.

(1)
```
    seqs_shape = dense_fsa_vec.shape  # we originally decoded with this
                                # dense_fsa_vec, it's the nnet outputs.
    # paths_shape will also be a k2.RaggedInt with 2 axes.
    paths_shape = k2.index(seqs_shape, path_to_seq_map);
```
`seqs_shape` is a `k2.RaggedShape` here. Do you want to convert it to a `k2.RaggedInt` by associating
a value`torch.arange(seqs_shape.row_ids(1).shape[0])` with it?

(2)
```
    with paths_shape.row_ids(1) as paths_shape_row_ids:
       pathframe_idx_to_seqframe_idx =  torch.arange(0, paths_shape_row_ids.shape[0]) - \
                     path_offsets[paths_shape.row_ids]
```

`paths_shape.row_ids(1)` is not a context manager here. Do you mean
```
paths_shape_row_ids = paths_shape.row_ids(1)
pathframe_idx_to_seqframe_idx =  torch.arange(0, paths_shape_row_ids.shape[0]) - \
                     path_offsets[paths_shape_row_ids]
```
?

**Comment by danpovey** [comment]
Regarding "seqs_shape is a k2.RaggedShape here. Do you want to convert it to a k2.RaggedInt by associating
a valuetorch.arange(seqs_shape.row_ids(1).shape[0]) with it?":
    .. no, you'll need to add an extra wrapping so we have (ragged shape indexed by tensor) available in k2.

RE (2): yes, I didn't realize that synntax was limited to context managers.

**Comment by csukuangfj** [comment]
Regarding `pathframe_idx_to_seqframe_idx`, since `k2.ragged.index` can output a second value `value_indexes`,
can we use it to replace `pathframe_idx_to_seqframe_idx`?

**Comment by danpovey** [comment]
I don't know what you mean.

**Comment by danpovey** [comment]
You mean when we do `paths_shape = k2.index(seqs_shape, path_to_seq_map);`?
That might be what is needed, yes.  You can check.

**Comment by csukuangfj** [comment]
What I meant is
```python
paths_shape, pathframe_idx_to_seqframe_idx = k2.ragged.index(seqs_shape, path_to_seq_map, need_value_indexes=True);

**Comment by csukuangfj** [comment]
BTW,  is the following a typo (switching the order of row_splits1 and row_splits2):
```python
    paths_lats_arc2path = paths_lats.arcs.shape.row_splits(1)[paths_lats.arcs.shape.row_splits(2)]
```

**Comment by danpovey** [comment]
? that's what I have right now, maybe I fixed it before.

**Comment by csukuangfj** [comment]
I believe the correct way is
```python
    paths_lats_arc2path = paths_lats.arcs.shape.row_splits(2)[paths_lats.arcs.shape.row_splits(1)]
```

**Comment by danpovey** [comment]
no, mine was right.  the dimension of the quantity is the dimension of the thing inside the brackets.    The values are the values of the thing outside the brackets.

**Comment by csukuangfj** [comment]
It could happen that
```python
paths_lats.arcs.shape.row_splits(2)[i] >= paths_lats.arcs.shape.row_splits(1).shape[0]
```
which means your approach will result in out-of-bound access error.

---
There are several places in the codebase using `row_splits2[row_splits1]`:

https://github.com/k2-fsa/k2/blob/ccd53662ae07e39cdfd106551ebc9f6721c2b363/k2/csrc/ragged_ops.cu#L1165

https://github.com/k2-fsa/k2/blob/ccd53662ae07e39cdfd106551ebc9f6721c2b363/k2/csrc/fsa_algo.cu#L173

https://github.com/k2-fsa/k2/blob/ccd53662ae07e39cdfd106551ebc9f6721c2b363/k2/csrc/fsa_algo.cu#L175

https://github.com/k2-fsa/k2/blob/ccd53662ae07e39cdfd106551ebc9f6721c2b363/k2/csrc/fsa.cu#L466

**Comment by danpovey** [comment]
Oh sorry, should be row_ids not row_splits.  Will fix.

**Comment by csukuangfj** [comment]
I think the only operation currently not supported is
```
    # Note: if the following indexing operation, sparse_matrix[1d tensor], does not exist,
    # we may have to create it.  
```

`torch.index_select` also support sparse tensors. For example, 

<img width="562" alt="Screen Shot 2021-02-03 at 7 31 47 PM" src="https://user-images.githubusercontent.com/5284924/106741342-8173c280-6656-11eb-9d92-c179db080557.png">


Can we use it? If so, I think all functions are available now.

**Comment by danpovey** [comment]
Cool!

**Comment by csukuangfj** [comment]
>      phone_fsas.pathphone_idx = torch.arange((0, phone_fsas.arcs.shape[0]), dtype=torch.int32)

Can you explain what the above code means? They are some errors in calling `torch.arange`.

Is `pathphone_idx` the same as `seqphone_idx` (mentioned in https://github.com/k2-fsa/snowfall/issues/96) ?

**Comment by danpovey** [comment]
yes I think I meant seqphone_idx.   The (0, shape), maybe should be 0, shape,

**Comment by danpovey** [comment]
i mean `0,1,2,...  arcs.shape[0]-1]`

**Comment by csukuangfj** [comment]
`arcs` is of type `_k2.RaggedArc` and `arcs.shape` is `_k2.RaggedShape`.

Do you mean `0, 1, 2, ..., arcs.row_ids(2).shape[0] - 1`, i.e., `0, 1, 2, ..., arcs.num_elements() - 1`?

**Comment by danpovey** [comment]
Yes, sorry.

**Comment by csukuangfj** [comment]
I have one doubt about `paths_decoding_graphs`:

From https://github.com/k2-fsa/k2/issues/641#issue-796593136

```
     # now compile decoding graphs corresponding to `phone_seqs` by constructing
     # fsas from them (remember we already have the final -1's!) and composing with ctc_topo.
      paths_decoding_graphs = ...
```

And from https://github.com/k2-fsa/k2/issues/641#issuecomment-770548577

```
     phone_fsas = k2.create_linear_fsas(phone_seqs)
     # add self loops
     phone_fsas = k2.add_self_loops(phone_fsas)
     # set an attribute called pathphone_idx, which corresponds to the arc-index in `phone_fsas` with
     # self-loops.  each phone has an index but there are blanks between them and at the start and end.
     phone_fsas.pathphone_idx = torch.arange((0, phone_fsas.arcs.shape[0]), dtype=torch.int32)
```

If I understand correctly, we cannot replace the statement `k2.add_epsilon_self_loops` with
`treat_epsilons_specially==True` while calling `k2.compose`. The introduction of epsilon self-loops here affects
the positions of phones in a sequence. Therefore, we have to use `treat_epsilons_specially==False` in `k2.compose` in this case.

Is it correct? @danpovey

**Comment by danpovey** [comment]
I think my idea was that `phone_fsas` would be composed, as in k2.compose(), with `ctc_topo`, and  we can do that compose with `treat_epsilons_specially=False` because `phone_fsas` has epsilon self-loops.  The pathphone_idx will be a bit more than twice as long as the num-phones, because the epsilons do get a position.  That's actually intentional.  What we get is that between each pair of phones, and at the start and end of the utterance, is an epsilon that gets its own `pathphone_idx` value.  That gives us a position where at most one phone can be inserted between each phone in the n-best list; it allows for the 2nd-pass utterances to be longer than the 1st-pass utterances.  (Later, we may figure out a way to insert more than one such position between each pair of phones.)

**Comment by csukuangfj** [comment]
>     num_phone_post_paths = num_phone_post[pathframe_idx_to_seqframe_idx]

Do you mean

```python
num_phone_post_paths = num_phone_post[pathframe_idx_to_seqframe_idx[paths_lats.pathframe_idx]]
```

@danpovey

**Comment by danpovey** [comment]
No, I think my idea was as I wrote in that comment, that the row-index of num_phone_post_paths is pathframe_idx.  If we later need a version of it where the row-index is the arc-index in paths_lats, we can get that later on.
